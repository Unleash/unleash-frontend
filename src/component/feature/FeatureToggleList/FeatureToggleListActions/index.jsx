import FeatureToggleListActions from "./FeatureToggleListActions";

export default FeatureToggleListActions;
