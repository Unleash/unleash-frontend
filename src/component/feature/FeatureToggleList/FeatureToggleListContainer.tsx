import { useFeatures } from 'hooks/api/getters/useFeatures/useFeatures';
import { FeatureSchema } from 'openapi';
import { FeatureToggleListTable } from './FeatureToggleListTable/FeatureToggleListTable';

const featuresPlaceholder: FeatureSchema[] = Array(15).fill({
    name: 'Name of the feature',
    description: 'Short description of the feature',
    type: '-',
    createdAt: new Date(2022, 1, 1),
    project: 'projectID',
});

export const FeatureToggleListContainer = () => {
    const { features = [], loading } = useFeatures();

    return (
        <FeatureToggleListTable
            data={loading ? featuresPlaceholder : features}
            isLoading={loading}
        />
    );
};
