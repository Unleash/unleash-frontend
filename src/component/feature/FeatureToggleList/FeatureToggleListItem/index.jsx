import FeatureToggleListItem from "./FeatureToggleListItem";

export default FeatureToggleListItem;
