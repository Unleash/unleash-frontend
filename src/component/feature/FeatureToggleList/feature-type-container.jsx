import { connect } from 'react-redux';
import Component from './feature-type-component';

const mapStateToProps = state => ({
    types: state.featureTypes.toJS(),
});

const FeatureType = connect(mapStateToProps)(Component);

export default FeatureType;
