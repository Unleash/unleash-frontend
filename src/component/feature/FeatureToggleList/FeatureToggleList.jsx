import { useContext } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { Button, IconButton, List, ListItem, Tooltip } from '@material-ui/core';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { Add } from '@material-ui/icons';

import FeatureToggleListItem from './FeatureToggleListItem';
import SearchField from '../../common/SearchField/SearchField';
import FeatureToggleListActions from './FeatureToggleListActions';
import ConditionallyRender from '../../common/ConditionallyRender/ConditionallyRender';
import PageContent from '../../common/PageContent/PageContent';
import HeaderTitle from '../../common/HeaderTitle';

import loadingFeatures from './loadingFeatures';

import { CREATE_FEATURE } from '../../providers/AccessProvider/permissions';

import AccessContext from '../../../contexts/AccessContext';

import { useStyles } from './styles';
import ListPlaceholder from '../../common/ListPlaceholder/ListPlaceholder';
import { getCreateTogglePath } from '../../../utils/route-path-helpers';
import { NAVIGATE_TO_CREATE_FEATURE } from '../../../testIds';
import { resolveFilteredProjectId } from '../../../hooks/useFeaturesFilter';

const FeatureToggleList = ({
    features,
    revive,
    archive,
    loading,
    flags,
    filter,
    setFilter,
    sort,
    setSort,
}) => {
    const { hasAccess } = useContext(AccessContext);
    const styles = useStyles();
    const smallScreen = useMediaQuery('(max-width:800px)');
    const mobileView = useMediaQuery('(max-width:600px)');

    const setFilterQuery = v => {
        const query = v && typeof v === 'string' ? v.trim() : '';
        setFilter(prev => ({ ...prev, query }));
    };

    const resolvedProjectId = resolveFilteredProjectId(filter);
    const createURL = getCreateTogglePath(resolvedProjectId, flags.E);

    const renderFeatures = () => {
        features.forEach(e => {
            e.reviveName = e.name;
        });

        if (loading) {
            return loadingFeatures.map(feature => (
                <FeatureToggleListItem
                    key={feature.name}
                    feature={feature}
                    revive={revive}
                    hasAccess={hasAccess}
                    className={'skeleton'}
                    flags={flags}
                />
            ));
        }

        return (
            <ConditionallyRender
                condition={features.length > 0}
                show={features.map(feature => (
                    <FeatureToggleListItem
                        key={feature.name}
                        feature={feature}
                        revive={revive}
                        hasAccess={hasAccess}
                        flags={flags}
                    />
                ))}
                elseShow={
                    <ConditionallyRender
                        condition={archive}
                        show={
                            <ListItem className={styles.emptyStateListItem}>
                                No archived features.
                            </ListItem>
                        }
                        elseShow={
                            <ListPlaceholder
                                text="No features available. Get started by adding a
                                new feature toggle."
                                link={createURL}
                                linkText="Add your first toggle"
                            />
                        }
                    />
                }
            />
        );
    };

    const headerTitle = archive ? 'Archived Features' : 'Features';

    return (
        <div className={styles.featureContainer}>
            <div className={styles.searchBarContainer}>
                <SearchField
                    initialValue={filter.query}
                    updateValue={setFilterQuery}
                    className={classnames(styles.searchBar, {
                        skeleton: loading,
                    })}
                />
                <ConditionallyRender
                    condition={!mobileView && !archive}
                    show={<Link to="/archive">Archive</Link>}
                />
            </div>

            <PageContent
                headerContent={
                    <HeaderTitle
                        loading={loading}
                        title={headerTitle}
                        actions={
                            <div className={styles.actionsContainer}>
                                <ConditionallyRender
                                    condition={!smallScreen}
                                    show={
                                        <FeatureToggleListActions
                                            filter={filter}
                                            setFilter={setFilter}
                                            sort={sort}
                                            setSort={setSort}
                                            loading={loading}
                                        />
                                    }
                                />
                                <ConditionallyRender
                                    condition={!archive}
                                    show={
                                        <ConditionallyRender
                                            condition={smallScreen}
                                            show={
                                                <Tooltip title="Create feature toggle">
                                                    <IconButton
                                                        component={Link}
                                                        to={createURL}
                                                        data-test={
                                                            NAVIGATE_TO_CREATE_FEATURE
                                                        }
                                                        disabled={
                                                            !hasAccess(
                                                                CREATE_FEATURE,
                                                                resolvedProjectId
                                                            )
                                                        }
                                                    >
                                                        <Add />
                                                    </IconButton>
                                                </Tooltip>
                                            }
                                            elseShow={
                                                <Button
                                                    to={createURL}
                                                    color="primary"
                                                    variant="contained"
                                                    component={Link}
                                                    data-test={
                                                        NAVIGATE_TO_CREATE_FEATURE
                                                    }
                                                    disabled={
                                                        !hasAccess(
                                                            CREATE_FEATURE,
                                                            resolvedProjectId
                                                        )
                                                    }
                                                    className={classnames({
                                                        skeleton: loading,
                                                    })}
                                                >
                                                    Create feature toggle
                                                </Button>
                                            }
                                        />
                                    }
                                />
                            </div>
                        }
                    />
                }
            >
                <List>{renderFeatures()}</List>
            </PageContent>
        </div>
    );
};

FeatureToggleList.propTypes = {
    features: PropTypes.array.isRequired,
    revive: PropTypes.func,
    loading: PropTypes.bool,
    archive: PropTypes.bool,
    flags: PropTypes.object,
    filter: PropTypes.object.isRequired,
    setFilter: PropTypes.func.isRequired,
    sort: PropTypes.object.isRequired,
    setSort: PropTypes.func.isRequired,
};

export default FeatureToggleList;
