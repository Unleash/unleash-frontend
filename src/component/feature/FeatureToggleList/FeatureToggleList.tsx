import { Dispatch, SetStateAction, useContext, VFC } from 'react';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { List, ListItem } from '@material-ui/core';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { IFlags } from 'interfaces/uiConfig';
import { SearchField } from 'component/common/SearchField/SearchField';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import PageContent from 'component/common/PageContent/PageContent';
import { HeaderTitle } from 'component/common/HeaderTitle/HeaderTitle';
import AccessContext from 'contexts/AccessContext';
import ListPlaceholder from 'component/common/ListPlaceholder/ListPlaceholder';
import { IFeaturesFilter } from 'hooks/useFeaturesFilter';
import { FeatureToggleListItem } from './FeatureToggleListItem/FeatureToggleListItem';
import { FeatureToggleListActions } from './FeatureToggleListActions/FeatureToggleListActions';
import { CreateFeatureButton } from '../CreateFeatureButton/CreateFeatureButton';
import { useCreateFeaturePath } from '../CreateFeatureButton/useCreateFeaturePath';
import { IFeaturesSort } from 'hooks/useFeaturesSort';
import { FeatureSchema } from 'openapi';
import { useStyles } from './styles';

interface IFeatureToggleListProps {
    features: FeatureSchema[];
    loading?: boolean;
    flags?: IFlags;
    filter: IFeaturesFilter;
    setFilter: Dispatch<SetStateAction<IFeaturesFilter>>;
    sort: IFeaturesSort;
    setSort: Dispatch<SetStateAction<IFeaturesSort>>;
    onRevive?: (feature: string) => void;
    inProject?: boolean;
    isArchive?: boolean;
}

const loadingFeaturesPlaceholder: FeatureSchema[] = Array(10)
    .fill({
        createdAt: '2021-03-19T09:16:21.329Z',
        description: ' ',
        enabled: true,
        lastSeenAt: '2021-03-24T10:46:38.036Z',
        name: '',
        project: 'default',
        stale: true,
        strategies: [],
        variants: [],
        type: 'release',
        archived: false,
        environments: [],
        impressionData: false,
    })
    .map((feature, index) => ({ ...feature, name: `${index}` })); // ID for React key

export const FeatureToggleList: VFC<IFeatureToggleListProps> = ({
    features,
    onRevive,
    inProject,
    isArchive,
    loading,
    flags,
    filter,
    setFilter,
    sort,
    setSort,
}) => {
    const { hasAccess } = useContext(AccessContext);
    const createFeature = useCreateFeaturePath(filter);
    const styles = useStyles();
    const smallScreen = useMediaQuery('(max-width:800px)');
    const mobileView = useMediaQuery('(max-width:600px)');

    const setFilterQuery = (v: string) => {
        const query = v && typeof v === 'string' ? v.trim() : '';
        setFilter(prev => ({ ...prev, query }));
    };

    const renderFeatures = () => {
        if (loading) {
            return loadingFeaturesPlaceholder.map(feature => (
                <FeatureToggleListItem
                    key={feature.name}
                    feature={feature}
                    onRevive={onRevive}
                    hasAccess={hasAccess}
                    className={'skeleton'}
                    flags={flags}
                />
            ));
        }

        return (
            <ConditionallyRender
                condition={features.length > 0}
                show={features.map(feature => (
                    <FeatureToggleListItem
                        key={feature.name}
                        feature={feature}
                        onRevive={onRevive}
                        hasAccess={hasAccess}
                        flags={flags}
                        inProject={inProject}
                    />
                ))}
                elseShow={
                    <ConditionallyRender
                        condition={Boolean(isArchive)}
                        show={
                            <ListItem className={styles.emptyStateListItem}>
                                No archived features.
                            </ListItem>
                        }
                        elseShow={
                            <ConditionallyRender
                                condition={Boolean(createFeature?.access)}
                                show={() => (
                                    <ListPlaceholder
                                        text="No features available. Get started by adding a new feature toggle."
                                        link={createFeature?.path}
                                        linkText="Add your first toggle"
                                    />
                                )}
                            />
                        }
                    />
                }
            />
        );
    };

    const searchResultsHeader = filter.query
        ? ` (${features.length} matches)`
        : '';

    const headerTitle = isArchive
        ? inProject
            ? `Project archived features ${searchResultsHeader}`
            : `Archived Features ${searchResultsHeader}`
        : `Features ${searchResultsHeader}`;

    return (
        <div>
            <div
                className={classnames(styles.searchBarContainer, {
                    dense: inProject,
                })}
            >
                <SearchField
                    initialValue={filter.query}
                    updateValue={setFilterQuery}
                    showValueChip={!mobileView}
                    className={classnames(styles.searchBar, {
                        skeleton: loading,
                    })}
                />
                <ConditionallyRender
                    condition={!mobileView && !isArchive}
                    show={<Link to="/archive">Archive</Link>}
                />
            </div>

            <PageContent
                headerContent={
                    <HeaderTitle
                        loading={loading}
                        title={headerTitle}
                        actions={
                            <div className={styles.actionsContainer}>
                                <ConditionallyRender
                                    condition={!smallScreen}
                                    show={
                                        <FeatureToggleListActions
                                            filter={filter}
                                            setFilter={setFilter}
                                            sort={sort}
                                            setSort={setSort}
                                            loading={loading}
                                            inProject={inProject}
                                        />
                                    }
                                />
                                <ConditionallyRender
                                    condition={!isArchive}
                                    show={
                                        <CreateFeatureButton
                                            filter={filter}
                                            loading={Boolean(loading)}
                                        />
                                    }
                                />
                            </div>
                        }
                    />
                }
            >
                <List>{renderFeatures()}</List>
            </PageContent>
        </div>
    );
};
