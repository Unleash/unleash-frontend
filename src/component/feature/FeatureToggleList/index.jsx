import { connect } from 'react-redux';
import { toggleFeature, fetchFeatureToggles } from '../../../store/feature-toggle/actions';
import { updateSettingForGroup } from '../../../store/settings/actions';
import FeatureToggleList from './FeatureToggleList';

import { hasPermission } from '../../../permissions';

function checkConstraints(strategy, regex) {
    if (!strategy.constraints) {
        return;
    }
    return strategy.constraints.some(c => c.values.some(v => regex.test(v)));
}

export const mapStateToPropsConfigurable = isFeature => state => {
    const featureMetrics = state.featureMetrics.toJS();
    const settings = state.settings.toJS().feature || {};
    let features = isFeature ? state.features.toJS() : state.archive.get('list').toArray();

    if (settings.currentProjectId && settings.currentProjectId !== '*') {
        features = features.filter(f => f.project === settings.currentProjectId);
    }

    if (settings.filter) {
        try {
            const regex = new RegExp(settings.filter, 'i');
            features = features.filter(
                feature =>
                    feature.strategies.some(s => checkConstraints(s, regex)) ||
                    regex.test(feature.name) ||
                    regex.test(feature.description) ||
                    feature.strategies.some(s => s && s.name && regex.test(s.name)) ||
                    (settings.filter.length > 1 && regex.test(JSON.stringify(feature)))
            );
        } catch (e) {
            // Invalid filter regex
        }
    }

    if (!settings.sort) {
        settings.sort = 'name';
    }

    if (settings.sort === 'enabled') {
        features = features.sort((a, b) =>
            // eslint-disable-next-line
            a.enabled === b.enabled ? 0 : a.enabled ? -1 : 1
        );
    } else if (settings.sort === 'stale') {
        features = features.sort((a, b) =>
            // eslint-disable-next-line
            a.stale === b.stale ? 0 : a.stale ? -1 : 1
        );
    } else if (settings.sort === 'created') {
        features = features.sort((a, b) => (new Date(a.createdAt) > new Date(b.createdAt) ? -1 : 1));
    } else if (settings.sort === 'Last seen') {
        features = features.sort((a, b) => (new Date(a.lastSeenAt) > new Date(b.lastSeenAt) ? -1 : 1));
    } else if (settings.sort === 'name') {
        features = features.sort((a, b) => {
            if (a.name < b.name) {
                return -1;
            }
            if (a.name > b.name) {
                return 1;
            }
            return 0;
        });
    } else if (settings.sort === 'strategies') {
        features = features.sort((a, b) => (a.strategies.length > b.strategies.length ? -1 : 1));
    } else if (settings.sort === 'type') {
        features = features.sort((a, b) => {
            if (a.type < b.type) {
                return -1;
            }
            if (a.type > b.type) {
                return 1;
            }
            return 0;
        });
    } else if (settings.sort === 'metrics') {
        const target = settings.showLastHour ? featureMetrics.lastHour : featureMetrics.lastMinute;

        features = features.sort((a, b) => {
            if (!target[a.name]) {
                return 1;
            }
            if (!target[b.name]) {
                return -1;
            }
            if (target[a.name].yes > target[b.name].yes) {
                return -1;
            }
            return 1;
        });
    }

    return {
        features,
        featureMetrics,
        settings,
        hasPermission: hasPermission.bind(null, state.user.get('profile')),
    };
};
const mapStateToProps = mapStateToPropsConfigurable(true);
const mapDispatchToProps = {
    toggleFeature,
    fetcher: () => fetchFeatureToggles(),
    updateSetting: updateSettingForGroup('feature'),
};

const FeatureToggleListContainer = connect(mapStateToProps, mapDispatchToProps)(FeatureToggleList);

export default FeatureToggleListContainer;
