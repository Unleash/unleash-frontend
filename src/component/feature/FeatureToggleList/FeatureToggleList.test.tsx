import { FeatureToggleList } from './FeatureToggleList';
import { render } from 'utils/testRenderer';
import { CREATE_FEATURE } from 'component/providers/AccessProvider/permissions';

jest.mock('component/common/ProjectSelect/ProjectSelect');

const mockFeatures = [
    {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        project: 'default',
        strategies: [
            {
                id: '0',
                name: 'gradualRolloutRandom',
                constraints: [],
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
        archived: false,
        environments: [],
        type: '',
        variants: [],
        impressionData: false,
    },
];

describe('FeatureToggleList', () => {
    it('renders correctly with one feature', () => {
        const { container } = render(
            <FeatureToggleList
                filter={{ project: '*' }}
                setFilter={jest.fn()}
                sort={{ type: 'name' }}
                setSort={jest.fn()}
                features={mockFeatures}
            />,
            { permissions: [{ permission: CREATE_FEATURE }] }
        );

        expect(container).toMatchSnapshot();
    });

    it('renders correctly with one feature without permissions', () => {
        const { container } = render(
            <FeatureToggleList
                filter={{ project: '*' }}
                setFilter={jest.fn()}
                sort={{ type: 'name' }}
                setSort={jest.fn()}
                features={mockFeatures}
            />
        );

        expect(container).toMatchSnapshot();
    });
});
