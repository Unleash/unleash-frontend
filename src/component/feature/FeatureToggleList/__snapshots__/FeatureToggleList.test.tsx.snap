// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureToggleList renders correctly with one feature 1`] = `
<div>
  <div>
    <div
      class="makeStyles-searchBarContainer-3"
    >
      <form
        class="makeStyles-container-6"
        role="search"
      >
        <div
          class="makeStyles-search-7 makeStyles-searchBar-4"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-searchIcon-8"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
          <div
            class="MuiInputBase-root makeStyles-inputRoot-9"
          >
            <input
              aria-label="search"
              class="MuiInputBase-input MuiInputBase-inputTypeSearch"
              placeholder="Search..."
              type="search"
              value=""
            />
          </div>
        </div>
      </form>
      <a
        href="/archive"
      >
        Archive
      </a>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      style="border-radius: 10px; box-shadow: none;"
    >
      <div
        class="makeStyles-headerContainer-10"
      >
        <div
          class="makeStyles-headerTitleContainer-14"
        >
          <div
            class=""
            data-loading="true"
          >
            <h1
              class="MuiTypography-root makeStyles-headerTitle-15 MuiTypography-h1"
            >
              Features 
            </h1>
          </div>
          <div
            class="makeStyles-headerActions-16"
          >
            <div
              class="makeStyles-actionsContainer-1"
            >
              <div
                class="makeStyles-actions-17"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2"
                  data-loading="true"
                >
                  Sorted by:
                </p>
                <button
                  aria-controls="sorting"
                  aria-haspopup="true"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text"
                  data-loading="true"
                  id="sorting"
                  style="text-transform: lowercase; font-weight: normal;"
                  tabindex="0"
                  title="Sort by"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    By Name
                    <span
                      class="MuiButton-endIcon MuiButton-iconSizeMedium"
                    >
                      <span
                        aria-hidden="true"
                        class="material-icons MuiIcon-root"
                      >
                        <svg
                          class="MuiSvgIcon-root"
                          focusable="false"
                          role="img"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                          <title>
                            Toggle
                          </title>
                        </svg>
                      </span>
                    </span>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="makeStyles-bodyContainer-11"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root makeStyles-listItem-18 MuiListItem-gutters"
          >
            <span
              class="makeStyles-listItemMetric-19"
            >
              <div
                class="makeStyles-container-25"
                style="background: rgb(237, 240, 241); font-size: 0.8rem;"
                title="No usage reported from connected applications"
              >
                <span
                  style="font-size: 1.4rem;"
                >
                  ⊕
                </span>
              </div>
            </span>
            <span
              class="makeStyles-listItemType-20 hideLt600"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-icon-26"
                data-loading="true"
                focusable="false"
                title="This is a \\"\\" toggle"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
                />
              </svg>
            </span>
            <span
              class="makeStyles-listItemLink-22"
            >
              <a
                class="listLink truncate"
                href="/projects/default/features/Another"
              >
                <span
                  class="toggleName"
                  title="another's description"
                >
                  Another
                   
                </span>
                <span />
                <small>
                  <time
                    datetime="2018-02-04T20:27:52.127Z"
                    title="2018-02-04T20:27:52.127Z"
                  >
                    4 years ago
                  </time>
                </small>
                <div>
                  <span
                    class="truncate"
                  >
                    <small>
                      another's description
                    </small>
                  </span>
                </div>
              </a>
            </span>
            <span
              class="makeStyles-listItemStrategies-23 hideLt920"
            >
              <a
                class="makeStyles-disabledLink-24"
                href="/projects/default"
                style="text-decoration: none;"
              >
                <div
                  class="MuiChip-root MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary"
                  style="margin-left: 8px; cursor: pointer;"
                  title="Project: default"
                >
                  <span
                    class="MuiChip-label"
                  >
                    default
                  </span>
                </div>
              </a>
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="polite"
    class="makeStyles-container-28"
    data-testid="ANNOUNCER_ELEMENT_TEST_ID"
    role="status"
  >
    Navigated to Features 
  </div>
</div>
`;

exports[`FeatureToggleList renders correctly with one feature without permissions 1`] = `
<div>
  <div>
    <div
      class="makeStyles-searchBarContainer-31"
    >
      <form
<<<<<<< HEAD
        class="makeStyles-container-34"
=======
        class="makeStyles-container-33"
>>>>>>> fix: update refectored components after rebase
        role="search"
      >
        <div
          class="makeStyles-search-35 makeStyles-searchBar-32"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-searchIcon-36"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
          <div
            class="MuiInputBase-root makeStyles-inputRoot-37"
          >
            <input
              aria-label="search"
              class="MuiInputBase-input MuiInputBase-inputTypeSearch"
              placeholder="Search..."
              type="search"
              value=""
            />
          </div>
        </div>
      </form>
      <a
        href="/archive"
      >
        Archive
      </a>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      style="border-radius: 10px; box-shadow: none;"
    >
      <div
        class="makeStyles-headerContainer-38"
      >
        <div
          class="makeStyles-headerTitleContainer-42"
        >
          <div
            class=""
            data-loading="true"
          >
            <h1
<<<<<<< HEAD
              class="MuiTypography-root makeStyles-headerTitle-43 MuiTypography-h1"
=======
              class="MuiTypography-root makeStyles-headerTitle-42 MuiTypography-h1"
>>>>>>> fix: update refectored components after rebase
            >
              Features 
            </h1>
          </div>
          <div
            class="makeStyles-headerActions-44"
          >
            <div
              class="makeStyles-actionsContainer-29"
            >
              <div
                class="makeStyles-actions-45"
              >
                <p
                  class="MuiTypography-root MuiTypography-body2"
                  data-loading="true"
                >
                  Sorted by:
                </p>
                <button
                  aria-controls="sorting"
                  aria-haspopup="true"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text"
                  data-loading="true"
                  id="sorting"
                  style="text-transform: lowercase; font-weight: normal;"
                  tabindex="0"
                  title="Sort by"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    By Name
                    <span
                      class="MuiButton-endIcon MuiButton-iconSizeMedium"
                    >
                      <span
                        aria-hidden="true"
                        class="material-icons MuiIcon-root"
                      >
                        <svg
                          class="MuiSvgIcon-root"
                          focusable="false"
                          role="img"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                          <title>
                            Toggle
                          </title>
                        </svg>
                      </span>
                    </span>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="makeStyles-bodyContainer-39"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root makeStyles-listItem-46 MuiListItem-gutters"
          >
            <span
              class="makeStyles-listItemMetric-47"
            >
              <div
                class="makeStyles-container-53"
                style="background: rgb(237, 240, 241); font-size: 0.8rem;"
                title="No usage reported from connected applications"
              >
                <span
                  style="font-size: 1.4rem;"
                >
                  ⊕
                </span>
              </div>
            </span>
            <span
              class="makeStyles-listItemType-48 hideLt600"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-icon-54"
                data-loading="true"
                focusable="false"
                title="This is a \\"\\" toggle"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
                />
              </svg>
            </span>
            <span
              class="makeStyles-listItemLink-50"
            >
              <a
                class="listLink truncate"
                href="/projects/default/features/Another"
              >
                <span
                  class="toggleName"
                  title="another's description"
                >
                  Another
                   
                </span>
                <span />
                <small>
                  <time
                    datetime="2018-02-04T20:27:52.127Z"
                    title="2018-02-04T20:27:52.127Z"
                  >
                    4 years ago
                  </time>
                </small>
                <div>
                  <span
                    class="truncate"
                  >
                    <small>
                      another's description
                    </small>
                  </span>
                </div>
              </a>
            </span>
            <span
              class="makeStyles-listItemStrategies-51 hideLt920"
            >
              <a
                class="makeStyles-disabledLink-52"
                href="/projects/default"
                style="text-decoration: none;"
              >
                <div
                  class="MuiChip-root MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary"
                  style="margin-left: 8px; cursor: pointer;"
                  title="Project: default"
                >
                  <span
                    class="MuiChip-label"
                  >
                    default
                  </span>
                </div>
              </a>
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="polite"
    class="makeStyles-container-56"
    data-testid="ANNOUNCER_ELEMENT_TEST_ID"
    role="status"
  >
    Navigated to Features 
  </div>
</div>
`;
